### ISSUE TITLE
Request feature: Customizable API_BASE

### ISSUE THREAD

**Issue by eMUQI** [machinewrapped/gpt-subtrans#47]
Hi,
I would like to request you to consider adding a new feature: allowing users to customize API_BASE. Since openai is not accessible in certain regions, using a proxy server can solve this issue. The openai python library already supports customizing API_BASE, which will enable us to connect to specific server addresses and make the application more flexible and convenient to use. Could you please add this feature to gpt-subtrans? Thank you very much!

**Comment by eMUQI** [comment]
My programming skills are limited, making it difficult for me to improve this project. However, I am currently able to use it by modifying the original code.

I have modified this file to set the apibase. https://github.com/machinewrapped/gpt-subtrans/blob/main/PySubtitleGPT/ChatGPTClient.py

I add `openai.api_base = "https://oa.api2d.net/v1"` , if this could become a configurable option, it would be great. I hope the author can consider this request. Thank you again for developing such an outstanding application.

```
def SendMessages(self, messages : list[str], temperature : float = None):
        """
        Make a request to the OpenAI API to provide a translation
        """
        options = self.options
        max_retries = options.get('max_retries', 3.0)
        backoff_time = options.get('backoff_time', 5.0)
        model = options.get('gpt_model')
        temperature = temperature or options.get('temperature', 0.0)

        translation = {}
        retries = 0

        while retries <= max_retries:
            try:
                openai.api_base = "https://oa.api2d.net/v1"
                response = openai.ChatCompletion.create(
                    model=model,
                    messages=messages,
                    temperature=temperature
                )

                translation['response_time'] = getattr(response, 'response_ms', 0)

                if response.usage:
                    translation['prompt_tokens'] = getattr(response.usage, 'prompt_tokens')
                    translation['completion_tokens'] = getattr(response.usage, 'completion_tokens')
                    translation['total_tokens'] = getattr(response.usage, 'total_tokens')

                # We only expect one choice to be returned as we have 0 temperature
                if response.choices:
                    choice = response.choices[0]
                    reply = response.choices[0].message

                    translation['finish_reason'] = getattr(choice, 'finish_reason', None)
                    translation['text'] = getattr(reply, 'content', None)
                else:
                    raise NoTranslationError("No choices returned in the response", response)

                # Return the response if the API call succeeds
                return translation
            
            except openai.error.RateLimitError as e:
                retry_after = e.headers.get('x-ratelimit-reset-requests') or e.headers.get('Retry-After')
                if retry_after:
                    retry_seconds = ParseDelayFromHeader(retry_after)
                    logging.warning(f"Rate limit hit, retrying in {retry_seconds} seconds...")
                    time.sleep(retry_seconds)
                    continue
                else:
                    logging.warning("Rate limit hit, quota exceeded. Please wait until the quota resets.")
                    raise

            except (openai.error.APIConnectionError, openai.error.Timeout, openai.error.ServiceUnavailableError) as e:
                if isinstance(e, openai.error.APIConnectionError) and not e.should_retry:
                    raise TranslationImpossibleError(str(e), translation)
                if retries == max_retries:
                    logging.warning(f"OpenAI failure {str(e)}, aborting after {retries} retries...")
                    raise
                else:
                    retries += 1
                    sleep_time = backoff_time * 2.0**retries
                    logging.warning(f"OpenAI error {str(e)}, retrying in {sleep_time}...")
                    time.sleep(sleep_time)
                    continue

            except Exception as e:
                raise TranslationImpossibleError(f"Unexpected error communicating with OpenAI", translation, error=e)

        return None

```

**Comment by machinewrapped** [comment]
Oh, that answers my question in the issue above! This should be quite straightforward to add, I'll see if I can get it done this week.

**Comment by machinewrapped** [comment]
Customisable base address is now available in v0.3.1 - let me know if you have any trouble!

**Comment by eMUQI** [comment]
When I try to run the GUI, I get the error shown in the screenshot. The window flashes briefly and it took me several attempts to capture it successfully. 
![image](https://github.com/machinewrapped/gpt-subtrans/assets/32332228/a8f52f50-215c-4430-857b-21428061f9e4)

**Comment by eMUQI** [comment]
`-b` option doesn't seem to be included in the latest release? How can I compile the executable file from source code? thanks
```
PS C:\Users\wulu\Downloads\gui-subtrans-0.3.1\gui-subtrans> .\gui-subtrans.exe --help
requests\__init__.py:102: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (5.1.0)/charset_normalizer (2.0.7) doesn't match a supported version!
usage: gui-subtrans.exe [-h] [-l TARGET_LANGUAGE] [-r RATELIMIT] [-k APIKEY] [-p PROJECT]
                        [--minbatchsize MINBATCHSIZE] [--maxbatchsize MAXBATCHSIZE] [--batchthreshold BATCHTHRESHOLD]
                        [--scenethreshold SCENETHRESHOLD] [--maxlines MAXLINES] [--theme THEME] [--firstrun]
                        [filepath]

Translates an SRT file using ChatGPT

positional arguments:
  filepath              Optional file to load on startup

options:
  -h, --help            show this help message and exit
  -l TARGET_LANGUAGE, --target_language TARGET_LANGUAGE
                        The target language for the translation
  -r RATELIMIT, --ratelimit RATELIMIT
                        Maximum number of batches per minute to process
  -k APIKEY, --apikey APIKEY
                        Your OpenAI API Key (https://platform.openai.com/account/api-keys)
  -p PROJECT, --project PROJECT
                        Read or Write project file to working directory
  --minbatchsize MINBATCHSIZE
                        Minimum number of lines to consider starting a new batch
  --maxbatchsize MAXBATCHSIZE
                        Maximum number of lines before starting a new batch is compulsory
  --batchthreshold BATCHTHRESHOLD
                        Number of seconds between lines to consider for batching
  --scenethreshold SCENETHRESHOLD
                        Number of seconds between lines to consider a new scene
  --maxlines MAXLINES   Maximum number of batches to process
  --theme THEME         Stylesheet to load
  --firstrun            Show the first-run options dialog on launch
Argument error: 0
```

**Comment by machinewrapped** [comment]
-b was only supported in the command line app (gpt-subtrans) since it's easy to configure it in the GUI settings. I've added it to the GUI command line arguments as well for consistency (gui-subtrans), and also to the first run options dialog.

I fixed that TypeError problem too! (v0.3.2)
